### PR TITLE
"Entities" Category for spawner

### PR DESCRIPTION
**Pull request by KittenCat180**
Still need to fix UI in some places (use Shift + a Menu for now)

### CODE REVIEW

**Review by TheDuckCow — COMMENTED** [review]
I want to move to using the term "entity" instead of models, to avoid confusion with meshswap (which is generally block specific models). Unless you want to make this change, I'm happy to push a commit to your fork which does this name update. Just know that on a git pull, you might end up with some conflicts, so again just let me know if it's faster for you to do this yourself.

would include renaming the file to entities.py, `def get_entities_cache`,  etc

File: MCprep_addon/mcprep_ui.py
Comment (TheDuckCow) [inline MCprep_addon/mcprep_ui.py:45]: Looks like this file was converted from tabs to spaces, creating a large diff. Hard to quickly see what changed, might be fastest if you push a commit that changes it back to tabs as indentation.

Comment (KittenCat1080ti) [reply MCprep_addon/mcprep_ui.py:45]: Ok, Will do about the indentation

I will also make the move to entities soon when I have time :) thanks

**Comment by KittenCat1080ti** [comment]
The feature has been completely implemented, All I need to do now is to rename everything from "models" to "entities" as you requested

File: MCprep_addon/mcprep_ui.py
Comment (TheDuckCow) [inline MCprep_addon/mcprep_ui.py:222]: fyi to generally never explicitly us "/" in path names, as they are OS specific. This would be properly written as: os.path.join(scriptdir, "MCprep_resources", "mcprep_models.blend)

File: MCprep_addon/mcprep_ui.py
Comment (TheDuckCow) [inline MCprep_addon/mcprep_ui.py:384]: nit pick: unnecessary white spaces/newlines. Optional to update, but generally I keep 1 newline between functions within a class (e.g. within a single operator definition), and 2 lines between the top-level modules (e.g. 2 lines between the end of one class to the next, or between functions defined on their own at the module level). Only exception is sometimes I use three between high level sections (e.g. where I separate operator definitions from registration; and technically isn't totally pythonic but I find improves readability)

File: MCprep_addon/spawner/spawn_util.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/spawn_util.py:112]: Can just delete this comment.

File: MCprep_addon/spawner/spawn_util.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/spawn_util.py:140]: Does this section need to be deleted, if just commented out? Or was this resolved in a more recent (not yet pushed?) update?

File: MCprep_addon/spawner/spawn_util.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/spawn_util.py:32]: nit: unnecessary new line between addon imports; newlines between "types" of imports is generally good though, e.g. between standard python imports, standard blender imports, and local addon imports.

File: MCprep_addon/spawner/spawn_util.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/spawn_util.py:229]: nit: Spaces vs tabs, surprised it doesn't complain but I suppose any newline/whitespace within lists is technically ok, but better to be consistent.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:91]: Optional refactor (ie if you don't, I'll do this after); move this function into `util.py`, and have both this module as well as the meshswap module reference this where used, to avoid duplicating the exact code.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:42]: FYI only, no action required for you: I am going to think how to refactor this so it's also in a way configurable and shared code with meshswap, noting it would have to be further abstracted since it specifically references the according entity or meshswap properties.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:209]: Question: does snapping make sense for this operator? Doesn't hurt to keep, but I'm thinking it's less relevant as entities are restricted to being placed like blocks are.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:339]: Optional refactor: Move this into `spawn_util.py`, and reference back here for both entities and the meshswap file's equivalent usage.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:380]: Optional refactor: ditto to above, move this into `spawn_util.py`, and reference back here for both entities and the meshswap file's equivalent usage.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:437]: Remember to update references here (and any other references in this file) to rename from Meshswap to Entities.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:449]: Okay to delete this commented line.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:461]: Should this operator be entirely deleted?

As I understand, the models should only be available as something you can "spawn" from the spawner, thus this "meshswap" type of auto placing it into the scene is not useful. Let me know if I am wrong and you were intended that it does replace entities somehow included in world imports.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:1286]: I believe this operator can be removed as well.

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:1328]: Remember to remove the operators here that are no longer needed (e.g. MCPREP_OT_fix_mineways_scale and MCPREP_OT_model)

File: MCprep_addon/spawner/models.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/models.py:1283]: (just calling out the is the endpoint of the _OT_models operator definition that I believe can be fully removed)

**Review by TheDuckCow — CHANGES_REQUESTED** [review]
Thanks for working on all of this! I realize I left a lot of comments, but as noted many are 'nits', and several are suggested refactors that I'm happy to do, more aimed at reducing code duplication.

One point of reflection, it might have been easier to start with a blank `models.py` (err, `entities.py`) file and paste in the copied code bits piece by piece, versus copying the whole file (which is why I think there is a lot of redundant pieces of code). Though I get that it makes getting it working earlier easier, makes it hard later to deduce "what isn't needed". Not a problem here as I have a good idea of what wasn't needed, but just good to note for contributions to code projects.

File: MCprep_addon/spawner/spawn_util.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/spawn_util.py:140]: While I was fixing bugs, I commented out some stuff, I still need to clean up. Thanks for reminding me

File: MCprep_addon/spawner/models.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/models.py:461]: I kind of wanted to implement it so that when meshswaping automatically, it would also replace things like minecarts. It would be an option in the meshswap options "Autorig entities" or something I was going to reference that class from the other meshswap file somehow. Probably not going to happen though. I should probably delete that

File: MCprep_addon/spawner/models.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/models.py:1328]: I was planning to use it to auto replace minecarts and such things. See a previous comment I made

File: MCprep_addon/spawner/models.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/models.py:1286]: See above comment I made

File: MCprep_addon/spawner/models.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/models.py:449]: Yes, will do. I need to remove all the comments that both me and you made

File: MCprep_addon/spawner/models.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/models.py:91]: Will do if I remember

File: MCprep_addon/spawner/models.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/models.py:209]: I agree, Will remove it

File: MCprep_addon/spawner/models.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/models.py]: Done

**Review by TheDuckCow — APPROVED** [review]
Thanks for making this change!

Note that I am approving this as-is, so the comments I've made are more optional "last round nits". After merging, I'm going to abstract out a couple more functions from the spawning operator so that the same code can be used both in the meshswapper and the entity spawner. Leading goal will be to reduce line-by-line duplications in the meshswap.py file as much as possible.

One more thing is, I'm planning to push out a "micro update" before I merge this in, just so you know! Likely ~about sometime this week.

File: MCprep_addon/spawner/meshswap.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/meshswap.py:236]: Deleted

File: MCprep_addon/spawner/spawn_util.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/spawn_util.py:138]: Yep

File: MCprep_addon/spawner/meshswap.py
Comment (KittenCat1080ti) [reply MCprep_addon/spawner/meshswap.py:91]: Done

**Review by TheDuckCow — APPROVED** [review]

**Comment by TheDuckCow** [comment]
Thanks for all this! As I mentioned, I'll merge in this pull request once I have my mini release out. One question in the meantime - are you able to share a snapshot of the entities.blend file?

**Comment by KittenCat1080ti** [comment]
I don't have an entities.blend file yet.. I am just using a blend file with a few differently scaled cubes to test. sorry.

**Comment by TheDuckCow** [comment]
Hey, just an FYI I'm going to be prioritizing to merge this into the v3.2.6 release, as per milestones below:
https://github.com/TheDuckCow/MCprep/milestone/5

Let me know if you want to be further involved, but by default I can take it from here. As an FYI, I'm likely going to pair this with a special contest to get contributions for the "entity" rigs and models.

**Comment by TheDuckCow** [comment]
A long time awaited, thank you for this contribution! I'm going to go ahead and merge this, and sort through any cleanup in the dev branch - this will be part of the end of year release in some fashion!

**Pull request merged**
